### PR TITLE
Add event based Olric metrics

### PR DESCRIPTION
**Pull request by kwapik**
### Description of change
This allows to track fragment migrate/received events count to monitor how Olric cluster is rebalanced.

##### Checklist

- [x] Tested in playground or other setup

## Summary by CodeRabbit

- **New Features**
  - Enhanced distributed cache metrics with new event tracking for fragment migrations and receptions.
  - Improved cache event monitoring capabilities with the integration of Olric event packages.
  - Introduced new context management for event listeners to ensure graceful shutdown.

- **Bug Fixes**
  - Implemented error handling in metrics reporting to increase system reliability.

- **Documentation**
  - Updated metric names for clarity in monitoring distributed cache events.

- **Refactor**
  - Refined `DistCache` structure with additional member identification fields to improve metrics accuracy.
  - Optimized metric scraping by introducing label extraction methods.

- **Chores**
  - Added utility functions for partition counting to aid in cache management tasks.

### CODE REVIEW

File: pkg/dist-cache/dist-cache.go [hunk 1]
```diff
@@ -145,6 +150,15 @@ func (dc *DistCache) scrapeMetrics(ctx context.Context) (proto.Message, error) {
 	return nil, nil
 }
 
+// getMetricsLabels returns metric labels based on dc.memberID and dc.memberName.
+// If any of those is empty, function will return false as second argument.
+func (dc *DistCache) getMetricsLabels() (prometheus.Labels, bool) {
+	metricLabels := make(prometheus.Labels)
+	metricLabels[metrics.DistCacheMemberIDLabel] = dc.memberID
+	metricLabels[metrics.DistCacheMemberNameLabel] = dc.memberName
+	return metricLabels, dc.memberID == "" || dc.memberName == ""
```
[the remark this answers is not in the document]
Comment (kwapik) [reply pkg/dist-cache/dist-cache.go:159]: Will fix.

File: pkg/dist-cache/provider.go [hunk 2]
```diff
@@ -180,9 +181,16 @@
...
+
+			err = dc.startReportingMetricsFromEvents(eventListenerCtx)
+			if err != nil {
+				log.Error().Err(err).Msg("Failed to start reporting metrics from events")
+				return err
+			}
 			return nil
 		},
 		OnStop: func(ctx context.Context) error {
+			eventListenerCancel()
```
[the remark this answers is not in the document]
Comment (kwapik) [reply pkg/dist-cache/provider.go:193]: ok

File: pkg/dist-cache/dist-cache-metrics.go [hunk 3]
```diff
@@ -9,14 +9,16 @@ import (
...
-	EntriesTotal          *prometheus.GaugeVec
-	DeleteHits            *prometheus.GaugeVec
-	DeleteMisses          *prometheus.GaugeVec
-	GetMisses             *prometheus.GaugeVec
-	GetHits               *prometheus.GaugeVec
-	EvictedTotal          *prometheus.GaugeVec
-	PartitionsCount       *prometheus.GaugeVec
-	BackupPartitionsCount *prometheus.GaugeVec
+	EntriesTotal                 *prometheus.GaugeVec
+	DeleteHits                   *prometheus.GaugeVec
```
Comment (tanveergill) [inline pkg/dist-cache/dist-cache-metrics.go:13]: Please change hits and musses to counters

File: pkg/dist-cache/dist-cache-metrics.go [hunk 4]
```diff
@@ -9,14 +9,16 @@ import (
...
+	DeleteMisses                 *prometheus.GaugeVec
+	GetMisses                    *prometheus.GaugeVec
+	GetHits                      *prometheus.GaugeVec
+	EvictedTotal                 *prometheus.GaugeVec
```
Comment (tanveergill) [inline pkg/dist-cache/dist-cache-metrics.go:17]: Evicted total is supposed to be counter

File: pkg/dist-cache/dist-cache-metrics.go (hunk 3 above)
Comment (tanveergill) [reply pkg/dist-cache/dist-cache-metrics.go:13]: Also, need separate result and global cache hit/miss counters

File: pkg/dist-cache/dist-cache-metrics.go (hunk 4 above)
Comment (kwapik) [reply pkg/dist-cache/dist-cache-metrics.go:17]: As disussed on a call, those are metrics from Olric stats, so they need to stay as is.

File: pkg/dist-cache/dist-cache-metrics.go (hunk 3 above)
Comment (kwapik) [reply pkg/dist-cache/dist-cache-metrics.go:13]: As disussed on a call, those are metrics from Olric stats, so they need to stay as is.

**Pull request merged**